File: src/main/java/xf/xflp/base/container/ZItemGraph.java
package xf.xflp.base.container;

import util.collection.LPListMap;
import util.collection.SetIndexArrayList;
import xf.xflp.base.item.PlacedItem;

import java.util.ArrayList;
import java.util.List;


/** 
 * Copyright (c) 2012-2026 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @author hschneid
 */
public class ZItemGraph {

	private final SetIndexArrayList<ZItemGraphEntry> lowerList = new SetIndexArrayList<>();
	private final SetIndexArrayList<List<PlacedItem>> upperList = new SetIndexArrayList<>();

	/**
	 * Adds a new item into Z-Graph. Due to linkage to internal data structure,
	 * the effort of adding is O(n+m)
	 *
	 */
	public void add(PlacedItem newItem, List<PlacedItem> itemList, LPListMap<Integer, Integer> zMap) {
		// Lower
		{
			List<PlacedItem> lowerItems = searchItemsBelow(newItem, itemList, zMap.get(newItem.z));
			ZItemGraphEntry e = new ZItemGraphEntry(newItem, lowerItems);
			lowerList.set(newItem.index, e);

			// Update of lower items with new upper item
			for (PlacedItem lowerItem : lowerItems) {
				if(upperList.get(lowerItem.index) == null)
					upperList.set(lowerItem.index, new ArrayList<>());
				upperList.get(lowerItem.index).add(newItem);
			}
		}

		// Upper
		{
			List<PlacedItem> upperItems = searchItemsAbove(newItem, itemList, zMap.get(newItem.zh));
			upperList.set(newItem.index, upperItems);

			// Update upper items with new lower item, which means new cut area ratio
			for (PlacedItem upperItem : upperItems) {
				ZItemGraphEntry e = lowerList.get(upperItem.index);
				e.lowerItemList.add(newItem);
				e.update();
			}
		}
	}

	/*
	 * Remove an item from Z graph
	 */
	public void remove(PlacedItem item) {
		// Remove Item from lower items
		if(lowerList.get(item.index) != null) {
			List<PlacedItem> get = lowerList.get(item.index).lowerItemList;
			for (int i = get.size() - 1; i >= 0; i--) {
				PlacedItem lowerItem = get.get(i);
				upperList.get(lowerItem.index).remove(item);
			}
		}

		// Entferne Item aus upper items
		if(upperList.get(item.index) != null) {
			List<PlacedItem> get = upperList.get(item.index);
			for (int i = get.size() - 1; i >= 0; i--) {
				PlacedItem upperItem = get.get(i);
				lowerList.get(upperItem.index).lowerItemList.remove(item);
			}
		}

		// Entferne lowerList-Eintrag
		lowerList.remove(item.index);
		// Entferne upperList-Eintrag
		upperList.remove(item.index);
	}

	public int size() {
		return upperList.size();
	}

	public List<PlacedItem> getItemsBelow(PlacedItem item) {
		return lowerList.get(item.index).lowerItemList;
	}

	public List<PlacedItem> getItemsAbove(PlacedItem item) {
		return upperList.get(item.index);
	}

	/**
	 * 
	 * @param item New item
	 * @param itemList List of all packed items
	 * @param zList List of all item indices where zh = item.z
	 * @return List of all items below new item
	 */
	private List<PlacedItem> searchItemsBelow(PlacedItem item, List<PlacedItem> itemList, List<Integer> zList) {
		List<PlacedItem> list = new ArrayList<>();

		if(item.z == 0)
			return list;

		for (int i = zList.size() - 1; i >= 0; i--) {
			int zItemIdx = zList.get(i);
			PlacedItem it = itemList.get(zItemIdx);

			if(it.zh == item.z &&
					it.xw > item.x && it.x < item.xw &&
					it.yl > item.y && it.y < item.yl)
				list.add(it);
		}

		return list;
	}

	private List<PlacedItem> searchItemsAbove(PlacedItem item, List<PlacedItem> itemList, List<Integer> zList) {
		List<PlacedItem> list = new ArrayList<>();

		if(zList == null)
			return list;

		for (int i = zList.size() - 1; i >= 0; i--) {
			int zItemIdx = zList.get(i);
			PlacedItem it = itemList.get(zItemIdx);

			if(it.z == item.zh &&
					it.xw > item.x && it.x < item.xw &&
					it.yl > item.y && it.y < item.yl)
				list.add(it);
		}

		return list;
	}
}

File: src/xflp/base/container/z_item_graph.py
"""
Z-Item graph for tracking vertical stacking relationships.

Copyright (c) 2012-2025 Holger Schneider
Licensed under the MIT License
"""

from typing import List, Optional, TYPE_CHECKING

from ...util.collection.set_index_array_list import SetIndexArrayList
from .z_item_graph_entry import ZItemGraphEntry

if TYPE_CHECKING:
    from ..item.item import Item
    from ...util.collection.lp_list_map import LPListMap


class ZItemGraph:
    """
    Graph structure tracking which items are stacked on top of each other.
    
    Maintains relationships between items along the Z-axis (height).
    For each item, tracks:
    - Which items are directly below it
    - Which items are directly above it
    
    This is used for load bearing calculations and stacking constraints.
    """
    
    def __init__(self):
        """Initialize an empty Z-item graph."""
        self._lower_list: SetIndexArrayList[ZItemGraphEntry] = SetIndexArrayList()
        self._upper_list: SetIndexArrayList[List['Item']] = SetIndexArrayList()
    
    def add(
        self, 
        new_item: 'Item', 
        item_list: List['Item'], 
        z_map: 'LPListMap[int, int]'
    ) -> None:
        """
        Add a new item into the Z-Graph.
        
        Links the item with items above and below it.
        
        Args:
            new_item: The item to add
            item_list: List of all items in container
            z_map: Map from Z-coordinate to item indices
        """
        # Process lower items
        lower_items = self._search_items_below(new_item, item_list, z_map.get(new_item.z))
        entry = ZItemGraphEntry(new_item, lower_items)
        self._lower_list.set(new_item.index, entry)
        
        # Update lower items with new upper item
        for lower_item in lower_items:
            if self._upper_list.get(lower_item.index) is None:
                self._upper_list.set(lower_item.index, [])
            self._upper_list.get(lower_item.index).append(new_item)
        
        # Process upper items
        upper_items = self._search_items_above(new_item, item_list, z_map.get(new_item.zh))
        self._upper_list.set(new_item.index, upper_items)
        
        # Update upper items with new lower item
        for upper_item in upper_items:
            entry = self._lower_list.get(upper_item.index)
            entry.lower_item_list.append(new_item)
            entry.update()
    
    def remove(self, item: 'Item') -> None:
        """
        Remove an item from the Z graph.
        
        Args:
            item: The item to remove
        """
        # Remove item from lower items
        if self._lower_list.get(item.index) is not None:
            lower_items = self._lower_list.get(item.index).lower_item_list
            for lower_item in lower_items:
                upper = self._upper_list.get(lower_item.index)
                if upper and item in upper:
                    upper.remove(item)
        
        # Remove item from upper items
        if self._upper_list.get(item.index) is not None:
            upper_items = self._upper_list.get(item.index)
            for upper_item in upper_items:
                entry = self._lower_list.get(upper_item.index)
                if entry and item in entry.lower_item_list:
                    entry.lower_item_list.remove(item)
        
        # Remove entries
        self._lower_list.remove(item.index)
        self._upper_list.remove(item.index)
    
    def size(self) -> int:
        """Get the number of items in the graph."""
        return self._upper_list.length()
    
    def get_items_below(self, item: 'Item') -> List['Item']:
        """
        Get all items directly below the given item.
        
        Args:
            item: The item to check
            
        Returns:
            List of items below
        """
        entry = self._lower_list.get(item.index)
        return entry.lower_item_list if entry else []
    
    def get_items_above(self, item: 'Item') -> List['Item']:
        """
        Get all items directly above the given item.
        
        Args:
            item: The item to check
            
        Returns:
            List of items above
        """
        items = self._upper_list.get(item.index)
        return items if items else []
    
    def _search_items_below(
        self, 
        item: 'Item', 
        item_list: List['Item'], 
        z_list: Optional[List[int]]
    ) -> List['Item']:
        """
        Find all items directly below the given item.
        
        Args:
            item: The item to check
            item_list: List of all items
            z_list: List of item indices at item.z height
            
        Returns:
            List of items below
        """
        result: List['Item'] = []
        
        if item.z == 0 or z_list is None:
            return result
        
        for z_item_idx in z_list:
            it = item_list[z_item_idx]
            if it is None:
                continue
                
            if (it.zh == item.z and
                it.xw > item.x and it.x < item.xw and
                it.yl > item.y and it.y < item.yl):
                result.append(it)
        
        return result
    
    def _search_items_above(
        self, 
        item: 'Item', 
        item_list: List['Item'], 
        z_list: Optional[List[int]]
    ) -> List['Item']:
        """
        Find all items directly above the given item.
        
        Args:
            item: The item to check
            item_list: List of all items
            z_list: List of item indices at item.zh height
            
        Returns:
            List of items above
        """
        result: List['Item'] = []
        
        if z_list is None:
            return result
        
        for z_item_idx in z_list:
            it = item_list[z_item_idx]
            if it is None:
                continue
                
            if (it.z == item.zh and
                it.xw > item.x and it.x < item.xw and
                it.yl > item.y and it.y < item.yl):
                result.append(it)
        
        return result
